exporter/tanzuobservabilityexporter/config.go

```diff
@@ -42,32 +44,123 @@ type Config struct {
 	exporterhelper.RetrySettings `mapstructure:"retry_on_failure"`
 
 	// Traces defines the Traces exporter specific configuration
-	Traces  TracesConfig  `mapstructure:"traces"`
-	Metrics MetricsConfig `mapstructure:"metrics"`
+	Traces  *TracesConfig  `mapstructure:"traces"`
+	Metrics *MetricsConfig `mapstructure:"metrics"`
 }
 
-func (c *Config) Validate() error {
-	tracesURL, err := parseEndpoint("traces", c.Traces.Endpoint)
+// withoutMissingFields fills in the Metrics or Traces field of this instance if it is missing.
+// withoutMissingFields returns a new Config instance with missing fields filled in while leaving c
+// unchanged.
+func (c *Config) withoutMissingFields() (filledIn Config, err error) {
+	if c.Traces == nil && c.Metrics == nil {
+		return Config{}, errors.New("either the traces or metrics stanza must be present")
+	}
+	filledIn = *c
+	var tracesURL, metricsURL *url.URL
+	if filledIn.Metrics == nil {
+		tracesURL, err = parseEndpoint("traces", filledIn.Traces.Endpoint)
+		if err != nil {
+			return Config{}, err
+		}
+		metricsURL = &url.URL{}
+		*metricsURL = *tracesURL
+		metricsURL.Host = hostWithPort(tracesURL.Host, 2878)
+		filledIn.Metrics = &MetricsConfig{HTTPClientSettings: confighttp.HTTPClientSettings{Endpoint: metricsURL.String()}}
+	} else if filledIn.Traces == nil {
+		metricsURL, err = parseEndpoint("metrics", filledIn.Metrics.Endpoint)
+		if err != nil {
+			return Config{}, err
+		}
+		tracesURL = &url.URL{}
+		*tracesURL = *metricsURL
+		tracesURL.Host = hostWithPort(metricsURL.Host, 30001)
```

Comment:
I'm a little confused.  If the user does not configure metrics, this is adding metrics configuration and using the supplied trace endpoint right? I think the part I'm missing is why the metrics settings need filled out in order for the exporter to work if it is only defined in a trace pipeline?  Or is the issue that users are adding the exporter to the metrics pipeline but not supplying metrics configuration?

Reply:
Good point. I think the intention (please correct me @keep94) was to be able to separately specify metrics configuration and traces configuration and when only one is available, assume it's like the other but on a different (default) port. 

Perhaps that's something tied to tanzuobservability but if not, I would suggest to consider approach [similar to used in OTLP/HTTP](https://github.com/open-telemetry/opentelemetry-collector/tree/main/exporter/otlphttpexporter#otlphttp-exporter). Either a single endpoint can be configured (in this case, this would be the hostname) or specifically `traces_endpoint/metrics_endpoint/logs_endpoint`. 

Also, if endpoint for the signal was not defined and it's used in the pipeline it should just trigger an error

Reply:
I haven't verified this yet, but perhaps the `createMetricsExporter` function only gets called when our exporter is used in a metrics pipeline. Perhaps all I have to do is throw an error in `createMetricsExporter` if there is no metrics configuration. But, I just remembered that for tanzuobservability we have a unique requirement.  Our traces exporter uses both the metrics and traces endpoint.  Before we added the metrics exporter, we hardcoded the metrics endpoint to be the same as the traces endpoint except use port 2878.  Now that we have both the metrics and traces exporter, we want the traces exporter to use the same metrics endpoint as the metrics exporter

Reply:
> Initially I wanted to throw an error if our tanzuobservability exporter is used in the metrics pipeline but no metrics configuration is defined, but I wasn't able to figure out how to do this in the code.

Unfortunately, `Validate()` can be only used in context of Config, without awareness of which pipeline it is. You might still throw an error when e.g. `createTraces` is called and not all config is available. 

Another option is that you might prefer hostname and port numbers. The port numbers would have default values, so they would always point somewhere. This has the limitation of not supporting url's with paths.

Looking at the ideas, I would circle back to how otlphttp exporter does it. Perhaps you could have three fields:

- `hostname` - if set, default port numbers are used and you can support both metrics and traces
- `metrics_endpoint` - by default: "https://\<hostname\>:2878" but could be overwritten here if needed
- `traces_endpoint` - be default: "https://\<hostname\>:30001" but could be overwritten here if needed

So a valid configuration would be:
```yaml
hostname: myserver.example.com
```

or:
```yaml
hostname: myserver.example.com
metrics_endpoint: https://myotherserver.example.com:2878
```

or:
```yaml
traces_endpoint: https://myserver1.example.com:30001
metrics_endpoint: https://myotherserver.example.com:2878
```

Reply:
I think this idea a lot.  I think it will simplify a lot of config.go, since you can be sure there is always an endpoint to use.

Reply:
Hi @pmm-sumo and @TylerHelmuth: I'm a maintainer of this exporter (in addition to @keep94), and I had question. I'll let Travis consider your suggestions and reply on his own.

Your config changes are backwards-incompatible with our existing yaml. Is there an otel-collector procedure or policy for introducing breaking changes like this? Whether or not we go down that path in this PR, it'd be good to know how to go about breaking changes in the future. I'm pretty sure I've seen other exporters change or deprecate config options, so I do think it's possible. Thanks!

Reply:
Hi @oppegard, I think this config change could be done in a non-breaking way. E.g. the new configuration fields could be added. When user uses the old one (essentially "endpoint"), a deprecation warning shows up in the logs.
After some time, the old configuration fields can be removed then.

Ultimately, I think it's your call since this is your exporter :)

Reply:
Example of deprecated config: 

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/b9915f749b7ec03fabcff3200bc659e3fe9f549d/processor/cumulativetodeltaprocessor/config.go#L35

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/b9915f749b7ec03fabcff3200bc659e3fe9f549d/processor/cumulativetodeltaprocessor/processor.go#L111

Agree with @pmm-sumo, its your exporter so you can handle this however you want.

Reply:
Thanks @pmm-sumo and @TylerHelmuth!